src/lib/notifier.c

```diff
@@ -14,82 +14,163 @@
 #include <sof/spinlock.h>
 #include <ipc/topology.h>
 
-static struct notify_data _notify_data __aligned(PLATFORM_DCACHE_ALIGN);
-
-void notifier_register(struct notifier *notifier)
+struct notify_data {
+	void *caller;
+	enum notify_id type;
+	uint32_t data_size;
+	void *data;
+} __aligned(PLATFORM_DCACHE_ALIGN);
+
+static struct notify_data _notify_data[PLATFORM_CORE_COUNT]
+	__aligned(PLATFORM_DCACHE_ALIGN);
+
+struct callback_handle {
+	void *receiver;
+	void *caller;
+	void (*cb)(void *arg, enum notify_id, void *data);
+	struct list_item list;
+};
+
+int notifier_register(void *receiver, void *caller, enum notify_id type,
+		       void (*cb)(void *arg, enum notify_id type, void *data))
 {
 	struct notify *notify = *arch_notify_get();
+	struct callback_handle *handle;
+
+	if (type >= NOTIFIER_ID_MAX)
+		return -EINVAL;
+
+	handle = rzalloc(RZONE_SYS_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*handle));
+
+	if (!handle)
+		return -ENOMEM;
+
+	list_init(&handle->list);
+	handle->receiver = receiver;
+	handle->caller = caller;
+	handle->cb = cb;
 
 	spin_lock(notify->lock);
-	list_item_prepend(&notifier->list, &notify->list);
+	list_item_prepend(&handle->list, &notify->list[type]);
 	spin_unlock(notify->lock);
+
+	return 0;
 }
 
-void notifier_unregister(struct notifier *notifier)
+void notifier_unregister(void *receiver, void *caller, enum notify_id type)
 {
 	struct notify *notify = *arch_notify_get();
-
+	struct list_item *wlist;
+	struct list_item *tlist;
+	struct callback_handle *handle;
+
+	if (type >= NOTIFIER_ID_MAX)
```

Comment:
Note there are static code analyzers that interpret enum as signed type and require to add also < <min_value> check.